sdks/python/apache_beam/io/gcp/gcsfilesystem.py

```diff
@@ -38,6 +40,15 @@ class GCSFileSystem(FileSystem):
   CHUNK_SIZE = gcsio.MAX_BATCH_OPERATION_SIZE  # Chuck size in batch operations
   GCS_PREFIX = 'gs://'
 
+  def __init__(self, pipeline_options):
+    super().__init__(pipeline_options)
+    if isinstance(pipeline_options, PipelineOptions):
+      gcs_options = pipeline_options.view_as(GoogleCloudOptions)
+      impersonate_service_account = gcs_options.impersonate_service_account
```

Comment:
I would suggest not putting specific impersonation logic in `gcsfilesystem`. I think what you want is some place to put a shared `get_credentials(gcp_options)` method that will handle doing whatever is necessary. Then the gcs filesystem and the dataflow runner can both use this. You just want to be careful that impersonation is used during job submission and not used during job execution. In practice for Java I did this by having them in the pipeline options but having them disappear when serialized and shipped to workers.

Reply:
I think I need to have it here, otherwise when the impersonated service tries to upload to a shared directory it won't be able to. Since credentials is a global variable it's important I set impersonation beforehand.

We can discuss a refactor, but for now, I can't get this to work otherwise.